Mathlib/SetTheory/ZFC/Ordinal.lean

```diff
@@ -120,17 +120,130 @@ theorem mem_trans (h : z.IsOrdinal) : x ∈ y → y ∈ z → x ∈ z :=
-  rintro ⟨h₁, ⟨h₂⟩⟩
-  use h₁
-  intro y z w hyz hzw hwx
-  have hzx := h₁.mem_trans hzw hwx
-  exact h₂ ⟨y, h₁.mem_trans hyz hzx⟩ ⟨z, hzx⟩ ⟨w, hwx⟩ hyz hzw
+theorem _root_.ZFSet.isOrdinal_iff_isTrans :
+    x.IsOrdinal ↔ x.IsTransitive ∧ IsTrans x.toSet (Subrel (· ∈ ·) _) where
+  mp h := ⟨h.isTransitive, h.isTrans⟩
+  mpr := by
+    rintro ⟨h₁, ⟨h₂⟩⟩
+    refine ⟨h₁, @fun y z w hyz hzw hwx ↦ ?_⟩
+    have hzx := h₁.mem_trans hzw hwx
+    exact h₂ ⟨y, h₁.mem_trans hyz hzx⟩ ⟨z, hzx⟩ ⟨w, hwx⟩ hyz hzw
+
+protected theorem mem (hx : x.IsOrdinal) (hy : y ∈ x) : y.IsOrdinal := by
+  have := hx.isTrans
+  let f : Subrel (· ∈ ·) y.toSet ↪r Subrel (· ∈ ·) x.toSet :=
+    Subrel.inclusionEmbedding (· ∈ ·) (hx.subset_of_mem hy)
+  exact isOrdinal_iff_isTrans.2 ⟨fun _ hz _ ha ↦ hx.mem_trans' ha hz hy, f.isTrans⟩
+
+/-- An ordinal is a transitive set of transitive sets. -/
+theorem _root_.ZFSet.isOrdinal_iff_forall_mem_isTransitive :
+    x.IsOrdinal ↔ x.IsTransitive ∧ ∀ y ∈ x, y.IsTransitive where
+  mp h := ⟨h.isTransitive, fun _ hy ↦ (h.mem hy).isTransitive⟩
+  mpr := fun ⟨h₁, h₂⟩ ↦ ⟨h₁, fun hyz hzw hwx ↦ (h₂ _ hwx).mem_trans hyz hzw⟩
+
+/-- An ordinal is a transitive set of ordinals. -/
+theorem _root_.ZFSet.isOrdinal_iff_forall_mem_isOrdinal :
+    x.IsOrdinal ↔ x.IsTransitive ∧ ∀ y ∈ x, y.IsOrdinal where
+  mp h := ⟨h.isTransitive, fun _ ↦ h.mem⟩
+  mpr := fun ⟨h₁, h₂⟩ ↦ isOrdinal_iff_forall_mem_isTransitive.2
+    ⟨h₁, fun y hy ↦ (h₂ y hy).isTransitive⟩
+
+theorem subset_iff_eq_or_mem (hx : x.IsOrdinal) (hy : y.IsOrdinal) : x ⊆ y ↔ x = y ∨ x ∈ y := by
+  constructor
+  · revert hx hy
+    apply Sym2.GameAdd.induction mem_wf _ x y
+    intro x y IH hx hy hxy
+    by_cases hyx : y ⊆ x
+    · exact Or.inl (subset_antisymm hxy hyx)
+    · obtain ⟨m, hm, hm'⟩ := mem_wf.has_min (y.toSet \ x.toSet) (Set.diff_nonempty.2 hyx)
+      have hmy : m ∈ y := show m ∈ y.toSet from Set.mem_of_mem_diff hm
+      have hmx : m ⊆ x := by
+        intro z hzm
+        by_contra hzx
+        exact hm' _ ⟨hy.mem_trans hzm hmy, hzx⟩ hzm
+      obtain rfl | H := IH m x (Sym2.GameAdd.fst_snd hmy) (hy.mem hmy) hx hmx
+      · exact Or.inr hmy
+      · cases Set.not_mem_of_mem_diff hm H
+  · rintro (rfl | h)
+    · rfl
+    · exact hy.subset_of_mem h
+
+alias ⟨eq_or_mem_of_subset, _⟩ := subset_iff_eq_or_mem
+
+theorem mem_of_subset_of_mem (h : x.IsOrdinal) (hz : z.IsOrdinal) (hx : x ⊆ y) (hy : y ∈ z) :
+    x ∈ z := by
+  obtain rfl | hx := h.eq_or_mem_of_subset (hz.mem hy) hx
+  · exact hy
+  · exact hz.mem_trans hx hy
+
+theorem not_mem_iff_subset (hx : x.IsOrdinal) (hy : y.IsOrdinal) : x ∉ y ↔ y ⊆ x := by
+  refine ⟨?_, fun hxy hyx ↦ mem_irrefl _ (hxy hyx)⟩
+  revert hx hy
+  apply Sym2.GameAdd.induction mem_wf _ x y
+  intros x y IH hx hy hyx z hzy
+  by_contra hzx
+  exact hyx (mem_of_subset_of_mem hx hy (IH z x (Sym2.GameAdd.fst_snd hzy) (hy.mem hzy) hx hzx) hzy)
+
+theorem not_subset_iff_mem (hx : x.IsOrdinal) (hy : y.IsOrdinal) : ¬ x ⊆ y ↔ y ∈ x := by
```

Comment:
Maybe we should define `instance : HasSSubset ZFSet` so you can state that ∈ is the same as ⊂.

Reply:
I like that idea. We're missing a few instances on `PSet` and `ZFSet` - most notably, we don't have the `Preorder` / `PartialOrder` structure defined. I'll take care of that in another PR.